design-documents/buildattr64-rationale.rst

```diff
@@ -0,0 +1,540 @@
+section. This would essentially be using GNU properties as a build attributes
+wrapper.
+
+From a modelling perspective GNU properties are present in both
+relocatable objects and loadable-units which influences the design of
+the properties with a simple combination rule per property for the
+static linker to transfer the relocatable-object properties to the
+loadable-unit. For example:
+
+- ``GNU_PROPERTY_STACK_SIZE`` has the static linker copy the maximum
+  value of all the relocatable object inputs
+  ``GNU_PROPERTY_STACK_SIZE`` to the output property.
+
+- ``GNU_PROPERTY_AARCH64_FEATURE_1_AND`` transfers the ``AND`` of all
+  the feature bits from the relocatable object inputs to the output
+  property. In practice an additional input for some the feature bits
+  has come from the static linker, making the implementation less
+  generic than it could be.
+
+The major differences between build attributes and GNU properties are:
+
+- Build attributes are relocatable object only and have an encoding
+  that is optimized for size.  The results of a build attributes
+  combination can be transferred to GNU properties or some platform
+  specific format in the loadable unit.
+
+- Build attributes have one combination operation ``join`` with the
+  partial order of the tag determining whether attributes accumulate
+  or diminish under ``join``.
+
+- There is a well defined partial order across a set of build
+  attributes. With compatibility represented by the ≤ operator.
+
+For independent binary attributes such as whether a particular
+architecture feature is required or not, build attributes offer no
+advantage over the feature bits provided in GNU properties. Build
+attributes can have advantages when modelling properties that:
+
+- can be represented by a value where the order of values represents
+  the degree to which the property has the value. For example if there
+  is a monotonically increasing set of hardware capabilities this can
+  be represented as a single attribute with a single value rather than
+  multiple feature bits with forcing functions. A concrete example is
+  ``FEAT_LSE2`` which can only exist if ``FEAT_LSE`` exists.
+
+- require more then one attribute to describe. For example the
+  (PAUTHABI64_) describes signing schema as a tuple of (vendor,
+  version). This would require two attributes to completely describe.
+
+Perhaps the strongest claim for build attributes is the ability to
+form a partial order between a two independent sets of build
+attributes.  This can be used as the basis of finding the best set of
+pre-compiled libraries out of a selection of pre-compiled
+alternatives. Intuitively the best candidate library makes the most
+demands of the execution environment without exceeding the
+capabilities of the exception environment.
+
+- All candidate libraries with build attributes ≤ selection build
+  attributes are compatible.
+
+- For each compatible candidate library L with best candidate B then
```

Comment:
I am puzzled by this sentence. Perhaps clarify it?

Reply:
OK, I'll try and rewrite it.